drivers/magnetometer/admt4000/admt4000.c

```diff
@@ -0,0 +1,2306 @@
+ * *******************************************************************************/
+int admt4000_ecc_encode(uint8_t *parity_num, uint8_t *code_length,
+                        uint8_t *code, uint8_t *input, uint8_t size_code, uint8_t size_input,
+                        uint8_t *ecc)
+{
+    int i = 0, j = 0, k = 0, n, m;
+    int eff_pos, value;
+    uint8_t position, xtract;
+
+    *ecc = 0; // initialize
+
+    n = size_code;
+    m = size_input;
+
+    if (n > 16 || m > 16)
+        return -EINVAL;
+
+    /* Compute parity bits needed */
+    *parity_num = 0;
+
+    while ((m * 8) > (1 << i) - (i + 1))
+    {
+        *parity_num += 1;
+        i++;
+    }
+
+    /* In bits */
+    *code_length = *parity_num + (m * 8);
+
+    for(i = 0; i < *code_length; i++)
+    {
+        /* Set to 0 */
+        if (i == ((1 << k) - 1))
+        {
+            code[(i / 8)] &= (uint8_t)~NO_OS_BIT((i % 8));
+
+            k++;
+        }
+        /* Get from input byte array */
+        else
+        {
+            code[(i / 8)] &= (uint8_t)~NO_OS_BIT((i % 8));
+
+            xtract = (uint8_t)no_os_field_get(NO_OS_BIT(j % 8), input[(j / 8)]);
+            xtract = (uint8_t)no_os_field_prep(NO_OS_BIT((i % 8)), xtract);
+
+            code[(i / 8)] |= xtract;
+
+            j++;
+        }
+    }
+
+    //Assignment of value bits inside the code array
+    /* parity_num is in BITS */
+    for (i = 0; i < *parity_num; i++)
+    {
+        position = 1 << i;
+
+        value = admt4000_hamming_calc(position, *code_length, code);
+
+        eff_pos = position - 1;
+
+        code[(eff_pos / 8)] &= ~NO_OS_BIT(eff_pos % 8);
+        code[(eff_pos / 8)] |= (NO_OS_BIT(eff_pos % 8) * value);
+
+        *ecc |= (value << i);
+    }
+
+    value = 0;
+    for (i = 0; i < *code_length; i++) {
+        uint8_t bit = no_os_field_get(NO_OS_BIT(i % 8), code[(i / 8)]);
+        if (bit)
+            value++;
+    }
+
+    if (value % 2)
+        value = 1;
+    else
+        value = 0;
+
+    *ecc |= (value << *parity_num);
+
+    return 0;
+}
+
+/***************************************************************************//**
+ * @brief Computes the Hamming Distance
+ *
+ * @param dev - The device structure.
+ * @param position - Bit position
+ * @param code_length - Length of code (original data + parity bits). Should be in bits
+ * @param code - Array containing bytes to iterate to
+ *
+ * @return 0 or 1 (to be appended during ECC encoding process)
+ * *******************************************************************************/
+int admt4000_hamming_calc(uint8_t position, uint8_t code_length, uint8_t *code)
+{
+    uint8_t bit_extract;
+    int count = 0, i, j;
+    i = position - 1;
+
+    if (NO_OS_ARRAY_SIZE(code) > 16)
+        return -EINVAL;
+
+    while (i < code_length)
+    {
+        for (j = i; j < i + position; j++)
+        {
+            bit_extract = no_os_field_get(NO_OS_BIT((j % 8)), code[(j / 8)]);
```

Comment:
see, this is exactly the kind of place where % operator should be optimized if possible, inefficiency is multiplied by the number of loops... so use the trick i told you about to optimize it